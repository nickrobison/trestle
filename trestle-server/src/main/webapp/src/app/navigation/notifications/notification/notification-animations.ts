import {animate, AnimationTriggerMetadata, state, style, transition, trigger} from "@angular/animations";

export const notificationAnimations: {
  readonly fadeNotification: AnimationTriggerMetadata,
  readonly buttonNotification: AnimationTriggerMetadata,
} = {
  fadeNotification: trigger('fadeAnimation', [
    state('in', style({opacity: 1})),
    transition('void => *', [style({
      opacity: 0.5
    }),
      animate('{{ fadeIn }}ms')]),
    transition('default => closing', animate('{{ fadeOut }}ms', style({ opacity: 0 })),
      ),
  ]),
  buttonNotification: trigger('buttonAnimation', [
    state('out', style({opacity: 0.5})),
    transition('in => out', [style({
      opacity: 0.5
    }),
    animate('500ms')]),
    transition('out => in', [style({
        opacity: 1
      }),
      animate('500ms')]),
  ]),
};

export type NotificationAnimationState = 'default' | 'closing';

export type NotificationButtonState = 'in' | 'out';
