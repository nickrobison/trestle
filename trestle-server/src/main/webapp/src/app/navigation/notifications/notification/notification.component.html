<div class="toast"
     [@fadeAnimation]="{value: animationState, params: {fadeIn: 1000, fadeOut: 1000}}"
     (@fadeAnimation.done)="onFadeFinished($event)"
>
  <div class="timer" [attr.data-status]="notification.state" style="{{'--duration: ' + notificationLifetime + ';'}}">
    <div class="background"></div>
    <div class="timebar"></div>
  </div>
  <div class="msg-text">
    {{getMessage()}}
  </div>
  <mat-icon (click)="triggerClose()"
            (mouseenter)="this.buttonState = 'in'"
            (mouseleave)="this.buttonState = 'out'"
            [@buttonAnimation]="{value: buttonState}">close
  </mat-icon>
</div>

