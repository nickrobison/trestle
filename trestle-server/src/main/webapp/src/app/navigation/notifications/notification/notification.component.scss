.toast {
  position: relative;
  display: flex;
  justify-content: space-around;
  margin: 0 10px 20px 0;
  padding: 10px 15px 10px 10px;
  width: 290px;
  background: #fff;
  border-width: 1px;
  border-style: solid;
  border-color: #dddddd #d6d6d6 #cfcfcf;
  box-shadow: 0 2px 4px rgba(0, 0, 0, 0.11);
}

.msg-text {
  margin-left: 5px;
}

$bar-width: 5px;
$opacity: 80%;

.timer .background {
  height: 100%;
  width: $bar-width;
  background: yellowgreen;
  position: absolute;
  left: 0;
  bottom: 0;
  opacity: $opacity;
}

.timer .timebar {
  height: 100%;
  width: $bar-width;
  background: green;
  position: absolute;
  left: 0;
  bottom: 0;
  opacity: $opacity;
  animation: roundtime calc(var(--duration) * 1s) linear forwards;
  transform-origin: bottom;
}

.timer[data-status="notification"] .background {
  background: yellowgreen;
}

.timer[data-status="notification"] .timebar {
  background: green;
}

.timer[data-status="error"] .background {
  background: indianred;
}

.timer[data-status="error"] .timebar {
  background: red;
}

@keyframes roundtime {
  to {
    transform: scaleY(0);
  }
}
