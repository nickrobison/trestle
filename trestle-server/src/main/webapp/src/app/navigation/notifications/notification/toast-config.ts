export class ToastData {
  msg: string;
}
