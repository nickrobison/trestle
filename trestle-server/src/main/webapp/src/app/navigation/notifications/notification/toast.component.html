<div class="toast">
  <p>toast works!</p>
</div>

