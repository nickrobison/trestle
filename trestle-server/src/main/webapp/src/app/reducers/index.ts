import {ActionReducerMap, MetaReducer} from '@ngrx/store';
import {storageMetaReducer} from './storage.metareducer';
import {authReducer, UserState} from "./auth.reducers";

export interface State {
  user: UserState;
}

export const reducers: ActionReducerMap<State> = {
  user: authReducer
};

export const metaReducers: MetaReducer<State>[] = [storageMetaReducer];
