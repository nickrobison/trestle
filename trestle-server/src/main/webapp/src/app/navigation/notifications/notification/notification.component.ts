import {Component, Input, OnDestroy, OnInit} from '@angular/core';
import {TrestleNotification} from "../../../reducers/notification.reducers";
import {Store} from "@ngrx/store";
import {State} from "../../../reducers";
import {removeNotification} from "../../../actions/notification.actions";
import {notificationAnimations, NotificationAnimationState, NotificationButtonState} from "./notification-animations";
import {AnimationEvent} from "@angular/animations";

@Component({
  selector: 'app-toast',
  templateUrl: './notification.component.html',
  styleUrls: ['./notification.component.scss'],
  animations: [notificationAnimations.fadeNotification, notificationAnimations.buttonNotification],
})
export class NotificationComponent implements OnInit, OnDestroy {

  @Input()
  notification: TrestleNotification;
  readonly notificationLifetime = 5;
  animationState: NotificationAnimationState = 'default';
  buttonState: NotificationButtonState = 'out';
  private intervalId: number;

  constructor(private store: Store<State>) {
  }

  ngOnInit(): void {
    this.intervalId = setTimeout(() => this.triggerClose(), this.notificationLifetime * 1000);
  }

  ngOnDestroy() {
    clearTimeout(this.intervalId);
  }

  onFadeFinished(event: AnimationEvent) {
    const {toState} = event;
    const isFadeOut = (toState as NotificationAnimationState) === 'closing';
    const itFinished = this.animationState === 'closing';
    if (itFinished && isFadeOut) {
      this.close();
    }
  }

  triggerClose() {
    this.animationState = 'closing';
  }

  private close() {
    this.store.dispatch(removeNotification({notification: this.notification}));
  }

  getMessage(): string {
    switch (this.notification.state) {
      case 'error':
        return this.notification.error.message;
      default:
        return this.notification.msg;
    }
  }

}
