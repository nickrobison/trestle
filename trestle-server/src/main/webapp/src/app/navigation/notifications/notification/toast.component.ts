import {Component, Input, OnDestroy, OnInit} from '@angular/core';
import {TrestleNotification} from "../../../reducers/notification.reducers";
import {Store} from "@ngrx/store";
import {State} from "../../../reducers";
import {removeNotification} from "../../../actions/notification.actions";

@Component({
  selector: 'app-toast',
  templateUrl: './toast.component.html',
  styleUrls: ['./toast.component.scss']
})
export class ToastComponent implements OnInit, OnDestroy {

  @Input()
  notification: TrestleNotification;
  private intervalId: number;

  constructor(private store: Store<State>) {
  }

  ngOnInit(): void {
    this.intervalId = setTimeout(() => this.store.dispatch(removeNotification({notification: this.notification})), 5000);
  }

  ngOnDestroy() {
    console.debug("Destroy called!");
    clearTimeout(this.intervalId);
  }

}
