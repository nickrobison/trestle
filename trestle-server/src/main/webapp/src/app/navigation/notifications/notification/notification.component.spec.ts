import {ComponentFixture, fakeAsync, TestBed, tick, waitForAsync} from '@angular/core/testing';

import {NotificationComponent} from './notification.component';
import {MaterialModule} from '../../../material/material.module';
import {MockStore, provideMockStore} from '@ngrx/store/testing';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import {removeNotification} from '../../../actions/notification.actions';
import {initialAppState} from '../../../reducers';

describe('NotificationComponent', () => {
  let component: NotificationComponent;
  let fixture: ComponentFixture<NotificationComponent>;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ NotificationComponent ],
      providers: [provideMockStore({
        initialState: initialAppState
      })],
      imports: [MaterialModule, NoopAnimationsModule],
    })
    .compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(NotificationComponent);
    component = fixture.componentInstance;
  });

  it('should be notification', () => {
    component.notification = {
      state: 'notification',
      msg: 'Hello there',
    };
    fixture.detectChanges();
    expect(component).toMatchSnapshot();
  });

  it('should be error', () => {
    component.notification = {
      state: 'error',
      error: new Error("I'm an error")
    };
    fixture.detectChanges();
    expect(component).toMatchSnapshot();
  });

  it('should dismiss', fakeAsync(() => {
    const triggerSpy = spyOn(component, 'triggerClose');
    component.notification = {
      state: 'notification',
      msg: 'Hello there',
    };
    fixture.detectChanges();
    tick(5000);
    expect(triggerSpy).toBeCalledTimes(1);
  }));
});
