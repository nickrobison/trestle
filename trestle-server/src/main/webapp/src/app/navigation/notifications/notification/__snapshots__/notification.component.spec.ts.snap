// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationComponent should be error 1`] = `
NotificationComponent {
  "animationState": "default",
  "buttonState": "out",
  "intervalId": 16,
  "notification": Object {
    "error": [Error: I'm an error],
    "state": "error",
  },
  "notificationLifetime": 5,
  "store": MockStore {
    "_isScalar": false,
    "actionsObserver": ActionsSubject {
      "_isScalar": false,
      "_value": Object {
        "type": "@ngrx/store/init",
      },
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "initialState": Object {
      "notifications": Object {
        "notifications": Array [],
      },
      "user": Object {
        "user": null,
        "userError": null,
        "userToken": "",
      },
    },
    "lastState": Object {
      "notifications": Object {
        "notifications": Array [],
      },
      "user": Object {
        "user": null,
        "userError": null,
        "userToken": "",
      },
    },
    "reducerManager": MockReducerManager {
      "_isScalar": false,
      "_value": [Function],
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "scannedActions$": Observable {
      "_isScalar": false,
      "source": ActionsSubject {
        "_isScalar": false,
        "_value": Object {
          "type": "@ngrx/store/init",
        },
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
    },
    "selectors": Map {},
    "source": MockState {
      "_isScalar": false,
      "_value": Object {
        "notifications": Object {
          "notifications": Array [],
        },
        "user": Object {
          "user": null,
          "userError": null,
          "userToken": "",
        },
      },
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "state$": MockState {
      "_isScalar": false,
      "_value": Object {
        "notifications": Object {
          "notifications": Array [],
        },
        "user": Object {
          "user": null,
          "userError": null,
          "userToken": "",
        },
      },
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
  },
}
`;

exports[`NotificationComponent should be notification 1`] = `
NotificationComponent {
  "animationState": "default",
  "buttonState": "out",
  "intervalId": 7,
  "notification": Object {
    "msg": "Hello there",
    "state": "notification",
  },
  "notificationLifetime": 5,
  "store": MockStore {
    "_isScalar": false,
    "actionsObserver": ActionsSubject {
      "_isScalar": false,
      "_value": Object {
        "type": "@ngrx/store/init",
      },
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "initialState": Object {
      "notifications": Object {
        "notifications": Array [],
      },
      "user": Object {
        "user": null,
        "userError": null,
        "userToken": "",
      },
    },
    "lastState": Object {
      "notifications": Object {
        "notifications": Array [],
      },
      "user": Object {
        "user": null,
        "userError": null,
        "userToken": "",
      },
    },
    "reducerManager": MockReducerManager {
      "_isScalar": false,
      "_value": [Function],
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "scannedActions$": Observable {
      "_isScalar": false,
      "source": ActionsSubject {
        "_isScalar": false,
        "_value": Object {
          "type": "@ngrx/store/init",
        },
        "closed": false,
        "hasError": false,
        "isStopped": false,
        "observers": Array [],
        "thrownError": null,
      },
    },
    "selectors": Map {},
    "source": MockState {
      "_isScalar": false,
      "_value": Object {
        "notifications": Object {
          "notifications": Array [],
        },
        "user": Object {
          "user": null,
          "userError": null,
          "userToken": "",
        },
      },
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
    "state$": MockState {
      "_isScalar": false,
      "_value": Object {
        "notifications": Object {
          "notifications": Array [],
        },
        "user": Object {
          "user": null,
          "userError": null,
          "userToken": "",
        },
      },
      "closed": false,
      "hasError": false,
      "isStopped": false,
      "observers": Array [],
      "thrownError": null,
    },
  },
}
`;
